Prevent styles from being added in the site editor

oandregal opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/55567

## What?

This PR fixes a bug by whiche the global styles are enqueued twice in the site editor.

## Why?

The site editor, unlike other editors, generates the styles coming from global styles (presets, styles) dynamically in the client. This is because the user can modify any value that we need to consolidate (merge core, theme, and user style preferences). We don't want to have the global styles twice, as it introduces subtle bugs.

## How?

This PR tags styles to be removed with the `isGlobalStyles: true` flag.

The styles sent by the server to the client end up in the `block-editor` store, to be found at `settings.styles`. By doing this, we leverage the existing mechanism in the site editor to recalculate the global styles upon user changes.

These are the styles before this PR:

![Captura de ecrã de 2022-04-12 14-28-20](https://user-images.githubusercontent.com/583546/162963190-c96f67e7-c8ca-4dd2-8c7b-579781b72bc5.png)

And after this PR:

![Captura de ecrã de 2022-04-12 14-25-46](https://user-images.githubusercontent.com/583546/162963163-416ed0fa-5e30-45f6-b45c-5c6f114bc31e.png)

Note how the first 3 styles are no longer in the store.

## Testing Instructions

Use this branch and verify that the block editor store has only 3 styles after this PR (would have 6 before).

Do some random manual test: add some block and set some preset values (colors, font sizes) as block styles and make sure they still work. Set some global styles and verify they work.

oandregal committed:
Prevent styles from being added in the site editor

oandregal commented on src/wp-includes/block-editor.php:
```diff
@@ -308,36 +308,37 @@ function get_block_editor_settings( array $custom_settings, $block_editor_contex
-		if ( '' !== $actual_css ) {
-			$preset_style['css'] = $actual_css;
-			$global_styles[]     = $preset_style;
+	if ( ! empty( $block_editor_context->post ) ) {
```
This would prevent styles from being added to any editor that doesn't have a post. However, we want those styles in all editors except for the site editor. How can we do this?

For context, Gutenberg does it [here](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/block-editor-settings.php#L15) (see `$context`).

youknowriad replied:
For me the ideal solution is to load these in all editors (no exceptions) but have a flag to know which ones are generated (global styles) and which ones are not.

oandregal replied:
Looking at the other PR https://github.com/WordPress/gutenberg/pull/40185 we do need two orthogonal things:

- be able to tell which are block styles and which are preset styles (the later should still be loaded when the user disables the "theme styles")
- be able to tell which styles come from "theme.json" and which don't (so we can overwrite the former in the site editor)

Load them all for all editors but mark them with the "styles source" solves the issue as well. Going to refactor this.

oandregal replied:
https://github.com/WordPress/wordpress-develop/pull/2542/commits/175b959c0896cf5a2328ccea56fbce2a37ef5628 updates the categories to be:

- Types: "presets" vs "styles" (was "theme" before)
- Sources: "theme_json" vs "css"

youknowriad replied:
This works for me 👍 I agree we need both things. Good things these are still unstable.

Do you think there's a way to detect the right type and source for how things were in WP 5.9 (mainly thinking about Gutenberg plugin compatibility here).

youknowriad replied:
I wonder if `theme_json` is a good value there. Technically it can come from variations, user global styles. so maybe "global styles" is the right value there, or something like it?

oandregal replied:
What do you think of this https://github.com/WordPress/gutenberg/pull/40185/files#r848389544 ? Instead of adding a new `__unstableSource` we can leverage the existing flag we use in the client `isGlobalStyles`.

oandregal committed:
Revert "Prevent styles from being added in the site editor"

This reverts commit f396ed824b7fa0db5d424e6ec977cd822306d75b.

oandregal committed:
Add __unstableSource and update __unstableType

We need to do two orthogonal things:

- Be able to tell which are block styles and which are preset styles:
  the later should still be loaded when the user disables the "theme styles".

- Be able to tell which styles come from "theme.json" and which don't:
  so we can overwrite the former in the site editor, for example.

oandregal committed:
Mark whether styles come from global styles

jorgefilipecosta commented:
Committed at https://github.com/WordPress/wordpress-develop/commit/2c2b5d0cfdd1036a57e8c3c50f0be32708546bfb.

jorgefilipecosta commented:
Committed at https://github.com/WordPress/wordpress-develop/commit/2c2b5d0cfdd1036a57e8c3c50f0be32708546bfb.
